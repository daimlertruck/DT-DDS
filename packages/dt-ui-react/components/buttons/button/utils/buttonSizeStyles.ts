import { CustomTheme as Theme } from '../../../../types';
import { ButtonSizes } from '../constants/types';

export const buttonSizeStyles = (size: ButtonSizes, theme: Theme) =>
  ({
    large: `
      ${theme.fontStyles.button1}
      padding: ${theme.spacing['2xs']} ${theme.spacing.s};
    `,
    medium: `
      ${theme.fontStyles.button2}
      padding: ${theme.spacing['3xs']} ${theme.spacing.xs};
    `,
    small: `
      ${theme.fontStyles.button2}
      padding: ${theme.spacing['4xs']} ${theme.spacing['2xs']};
    `,
  }[size]);
