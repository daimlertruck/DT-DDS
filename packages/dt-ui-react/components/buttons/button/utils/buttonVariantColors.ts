import { CustomTheme as Theme } from '../../../../types';
import {
  ButtonVariant,
  ContainedColors,
  OutlinedColors,
  TextColors,
} from '../constants/types';

interface ColorKeys {
  color: string;
  bgColor: string;
  hoverBgColor?: string;
  hoverColor?: string;
}

export const buttonVariantColors = (
  color: ContainedColors | OutlinedColors,
  theme: Theme
): Record<ButtonVariant, ColorKeys> => ({
  contained: {
    color: theme.colors.alpha.white,
    bgColor: theme.palette.alpha[color],
    hoverBgColor: theme.palette.alpha[`${color}Dark_300` as ContainedColors],
  },
  outlined: {
    color:
      color === 'neutral'
        ? theme.colors.alpha.gray_200
        : theme.palette.alpha[color],
    bgColor: theme.colors.alpha.white,
    hoverBgColor: theme.palette.alpha[`${color}Light_300` as OutlinedColors],
  },
  text: {
    color: theme.palette.alpha[color],
    bgColor: 'transparent',
    hoverColor: theme.palette.alpha[`${color as TextColors}Dark_300`],
  },
});
