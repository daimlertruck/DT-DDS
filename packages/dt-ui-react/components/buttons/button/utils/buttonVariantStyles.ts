import { CustomTheme as Theme } from '../../../../types';
import { ButtonVariant, OutlinedColors, SolidColors } from '../constants/types';

export const buttonVariantStyles = (
  variant: ButtonVariant,
  theme: Theme,
  color: SolidColors | OutlinedColors
) =>
  ({
    solid: `
      text-transform: uppercase;
      border: ${theme.spacing.none};

      &:focus {
        box-shadow: 0px 2px 1px 0px #1A1A1A0D, 0px 0px 0px 4px ${theme.colors.blue_10};
        background-color: ${theme.palette[color]};
      }

      &:disabled {
        opacity: 38%;
      }
  `,
    outlined: `
      border: 1px solid;
      text-transform: uppercase;
        ${
          color === 'neutral'
            ? theme.palette.neutralDark_100
            : theme.palette[color]
        };

      &:disabled {
        color: ${
          color === 'neutral'
            ? theme.palette.neutralLight_100
            : theme.palette[color]
        };
        background-color: ${theme.colors.white};
        border-color: ${
          color === 'neutral'
            ? theme.palette.neutralLight_100
            : theme.palette[color]
        };
        cursor: default;
      }
    `,
    text: `
      border: none;

      &:disabled {
        color: ${theme.palette.neutralLight_100};
      }
    `,
  }[variant]);
