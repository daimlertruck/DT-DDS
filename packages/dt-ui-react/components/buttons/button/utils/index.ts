export * from './buttonVariantColors';
export * from './buttonVariantStyles';
