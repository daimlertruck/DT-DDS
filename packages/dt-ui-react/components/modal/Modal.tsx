import { Backdrop, BackdropProps } from '@dt-ui/react-backdrop';
import { BaseProps, useClickOutside, Portal } from '@dt-ui/react-core';
import { forwardRef, RefObject, useRef } from 'react';

import { CloseIcon } from '../../core/assets';
import { IconButton } from '../buttons';
import { Spinner } from '../spinner';
import { Typography } from '../typography';

import {
  ContentStyled,
  ModalStyled,
  HeaderStyled,
  FooterStyled,
  HeaderWrapperStyled,
  ModalLoadingOverlay,
} from './Modal.styled';

export interface ModalProps extends BaseProps {
  handleClose: () => void;
  hasClickOutside?: boolean;
  isLoading?: boolean;
}
export interface ModalHeaderProps extends ModalProps {
  hasBackgroundColor?: boolean;
  hasBorder?: boolean;
  title: string;
  subTitle?: string;
  closeIconLabel?: string;
}

export interface ModalFooterProps extends BaseProps {
  hasBackgroundColor?: boolean;
  hasBorder?: boolean;
}

export const Overlay = forwardRef<HTMLDivElement, BackdropProps>(
  ({ children, dataTestId, isOpen, type }, ref) => {
    return (
      <Portal isOpen={isOpen}>
        <Backdrop dataTestId={dataTestId} isOpen={isOpen} ref={ref} type={type}>
          {children}
        </Backdrop>
      </Portal>
    );
  }
);
export const Modal = ({
  children,
  handleClose,
  hasClickOutside = false,
  isLoading = false,
}: ModalProps) => {
  const ref = useRef(null);

  useClickOutside({
    ref: hasClickOutside && !isLoading ? ref : ({} as RefObject<Element>),
    handler: handleClose,
  });

  return (
    <ModalStyled ref={ref}>
      {isLoading ? (
        <ModalLoadingOverlay>
          <Spinner size='extra-large' />
        </ModalLoadingOverlay>
      ) : null}
      {children}
    </ModalStyled>
  );
};

Modal.Header = ({
  dataTestId,
  handleClose,
  hasBackgroundColor = false,
  hasBorder = false,
  title,
  subTitle,
}: ModalHeaderProps) => {
  return (
    <HeaderStyled
      data-testid={dataTestId ?? 'modal-header'}
      hasBackgroundColor={hasBackgroundColor}
      hasBorder={hasBorder}
    >
      <HeaderWrapperStyled>
        <Typography color='content.heading' element='h4' fontStyles='h4'>
          {title}
        </Typography>

        <IconButton
          ariaLabel='close modal'
          color='default'
          onMouseDown={handleClose}
        >
          <CloseIcon />
        </IconButton>
      </HeaderWrapperStyled>
      {subTitle ? (
        <Typography color='content.body' element='h5' fontStyles='body1'>
          {subTitle}
        </Typography>
      ) : null}
    </HeaderStyled>
  );
};

Modal.Content = ({ children }: BaseProps) => {
  return <ContentStyled>{children}</ContentStyled>;
};

Modal.Footer = ({
  children,
  style,
  hasBackgroundColor = false,
  hasBorder = false,
}: ModalFooterProps) => {
  return (
    <FooterStyled
      hasBackgroundColor={hasBackgroundColor}
      hasBorder={hasBorder}
      style={style}
    >
      {children}
    </FooterStyled>
  );
};
