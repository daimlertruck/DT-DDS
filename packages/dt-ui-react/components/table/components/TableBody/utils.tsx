import { Children, ReactNode, cloneElement, isValidElement } from 'react';

import { TableDataCellHeader } from '../TableDataCell/TableDataCell';

export const addHeadersToDataCells = (
  dataCells: ReactNode,
  headers: ReactNode[]
) => {
  return Children.map(dataCells, (dataCell, index) => {
    if (isValidElement(dataCell)) {
      const header = headers[index];
      return cloneElement(
        dataCell,
        {},
        <>
          <TableDataCellHeader>{header}</TableDataCellHeader>
          {dataCell.props.children}
        </>
      );
    }
    return dataCell;
  });
};

export const addHeadersToDataCellsInRows = (
  rows: ReactNode,
  headers: ReactNode[]
) => {
  return Children.map(rows, (row) => {
    if (isValidElement(row)) {
      return cloneElement(
        row,
        {},
        addHeadersToDataCells(row.props.children, headers)
      );
    }
    return row;
  });
};
