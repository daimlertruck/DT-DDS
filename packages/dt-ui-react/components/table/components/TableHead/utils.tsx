import { Children, ReactNode, isValidElement } from 'react';

const extractColumnHeaders = (columnHeaders: ReactNode): ReactNode[] => {
  return Children.toArray(columnHeaders).reduce<ReactNode[]>(
    (headers, columnHeader) => {
      if (isValidElement(columnHeader)) {
        return [...headers, columnHeader.props.children];
      }

      return headers;
    },
    []
  );
};

export const extractTableHeaders = (theadRow: ReactNode) => {
  if (isValidElement(theadRow)) {
    return extractColumnHeaders(theadRow.props.children);
  }

  return [];
};
