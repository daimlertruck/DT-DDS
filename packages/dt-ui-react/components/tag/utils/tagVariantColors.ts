import { CustomTheme as Theme } from '../../../types';
import { TagColor, TagVariant } from '../constants';

interface TagVariantColors {
  border: string | number;
  color: string;
  borderColor?: string;
  backgroundColor: string;
  '&:hover'?: {
    borderColor?: string;
    backgroundColor: string;
    color: string;
  };
}

interface TagVariantColorsProps {
  theme: Theme;
  color: TagColor;
  variant: TagVariant;
  hasHover: boolean;
  isDisabled: boolean;
}

const getColorWithShade = ({
  theme,
  color,
  shade,
}: Pick<TagVariantColorsProps, 'color'> & {
  shade: '90' | '60' | '30' | '10';
  theme: Theme;
}) => {
  const { colors, palette } = theme;

  if (color === 'primary') {
    return {
      '90': palette.primaryDark_500,
      '60': palette.primaryDark_200,
      '30': palette.primaryLight_100,
      '10': palette.primaryLight_400,
    }[shade];
  }

  const colorWithShade: keyof Theme['colors'] = `${color}_${shade}`;

  return colors[colorWithShade];
};

export const tagVariantColors = ({
  theme,
  color,
  variant,
  hasHover,
  isDisabled,
}: TagVariantColorsProps): TagVariantColors => {
  const colorShade90 = getColorWithShade({
    theme,
    color,
    shade: '90',
  });
  const colorShade60 = getColorWithShade({
    theme,
    color,
    shade: '60',
  });
  const colorShade30 = getColorWithShade({
    theme,
    color,
    shade: '30',
  });
  const colorShade10 = getColorWithShade({
    theme,
    color,
    shade: '10',
  });

  const colorWithDisabled = isDisabled ? colorShade30 : colorShade60;

  return {
    solid: {
      border: theme.spacing.none,
      color: theme.colors.white,
      backgroundColor: colorWithDisabled,
      ...(hasHover && {
        '&:hover': {
          backgroundColor: colorShade90,
          color: theme.colors.white,
        },
      }),
    },
    outlined: {
      border: `1px solid ${colorShade60}`,
      color: colorWithDisabled,
      backgroundColor: colorShade10,
      borderColor: colorWithDisabled,
      ...(hasHover && {
        '&:hover': {
          borderColor: colorShade90,
          backgroundColor: colorShade10,
          color: colorShade90,
        },
      }),
    },
  }[variant];
};
