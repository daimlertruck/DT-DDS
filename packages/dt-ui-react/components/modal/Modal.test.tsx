import { fireEvent, render, screen, within } from '@testing-library/react';
import React from 'react';

import { withProviders } from '../../utils';
import { Button } from '../buttons';

import { Modal, Overlay } from './Modal';

describe('<Modal/> Component', () => {
  const ProvidedOverlay = withProviders(Overlay);

  const handleClick = jest.fn();

  describe('handle modal actions', () => {
    beforeEach(() => {
      render(
        <ProvidedOverlay isOpen={true} dataTestId='modal-overlay'>
          <Modal handleClose={handleClick}>
            <Modal.Header handleClose={handleClick}>Title</Modal.Header>
            <Modal.Content>
              <Modal.ContentTitle>Subtitle</Modal.ContentTitle>
              Some text here
            </Modal.Content>
            <Modal.Footer>
              <Button
                variant='outlined'
                color='neutral'
                onClick={handleClick}
                dataTestId='cancel'
              >
                Cancel
              </Button>
              <Button onClick={handleClick} dataTestId='action'>
                Action
              </Button>
            </Modal.Footer>
          </Modal>
        </ProvidedOverlay>
      );
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('renders modal correctly', () => {
      expect(screen.getByTestId('modal-overlay')).toMatchSnapshot();
    });

    it('fires the mock close function when clicking on the x', () => {
      const closeBtn = within(screen.getByTestId('modal-header')).getByRole(
        'button'
      );
      fireEvent.mouseDown(closeBtn);

      expect(handleClick).toBeCalledTimes(1);
    });

    it('fires the mock close function when clicking on cancel', () => {
      const cancelBtn = screen.getByTestId('cancel-button');
      fireEvent.click(cancelBtn);

      expect(handleClick).toBeCalledTimes(1);
    });

    it('fires the mock action function when clicking on action', () => {
      const actionBtn = screen.getByTestId('action-button');
      fireEvent.click(actionBtn);

      expect(handleClick).toBeCalledTimes(1);
    });
  });

  describe('handle close outside modal', () => {
    beforeEach(() => {
      jest
        .spyOn(React, 'useRef')
        .mockReturnValueOnce({ current: document.createElement('button') });
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('should fire handle close function when clicking outside', () => {
      render(
        <>
          <ProvidedOverlay isOpen={true} dataTestId='modal-overlay'>
            <Modal hasClickOutside={true} handleClose={handleClick}>
              Content
            </Modal>
          </ProvidedOverlay>
          <div>outside</div>
        </>
      );

      fireEvent.mouseDown(screen.getAllByText('outside')[0]);

      expect(handleClick).toBeCalledTimes(1);
    });

    it('should not fire handle close function when clicking outside', () => {
      render(
        <>
          <ProvidedOverlay isOpen={true} dataTestId='modal-overlay'>
            <Modal hasClickOutside={false} handleClose={handleClick}>
              Content
            </Modal>
          </ProvidedOverlay>
          <div>outside</div>
        </>
      );

      fireEvent.mouseDown(screen.getAllByText('outside')[0]);

      expect(handleClick).toBeCalledTimes(0);
    });
  });

  describe('handle disabled prop', () => {
    beforeEach(() => {
      render(
        <ProvidedOverlay isOpen={true} dataTestId='modal-overlay'>
          <Modal handleClose={handleClick}>
            <Modal.Header handleClose={handleClick} isDisabled={true}>
              Title
            </Modal.Header>
            <Modal.Content>Some text here</Modal.Content>
          </Modal>
        </ProvidedOverlay>
      );
    });

    it('should not be possible to click on the Close icon', () => {
      const closeBtn = within(screen.getByTestId('modal-header')).getByRole(
        'button'
      );

      expect(closeBtn).toBeDisabled();
    });
  });
});
