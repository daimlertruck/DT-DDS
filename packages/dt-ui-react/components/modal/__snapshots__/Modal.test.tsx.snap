// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal/> Component handle modal actions renders modal correctly 1`] = `
.emotion-0 {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 5;
  background: rgba(21,22,23,0.6);
}

.emotion-2 {
  width: 50%;
  max-width: 825px;
  position: fixed;
  z-index: 10;
  left: 50%;
  top: 50%;
  -webkit-transform: translate(-50%, -50%);
  -moz-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  background-color: white;
}

.emotion-4 {
  font-weight: 500;
  font-size: 1rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  font-weight: 700;
  color: #151617;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 32px;
  margin-bottom: 0;
  padding-bottom: 16px;
  border-bottom: 1px solid #323232;
  text-transform: uppercase;
}

.emotion-4 svg {
  cursor: pointer;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  cursor: pointer;
  background-color: transparent;
  color: #4F4F4F;
}

.emotion-6:hover {
  color: #D3D4D5;
}

.emotion-8 {
  padding: 32px;
}

.emotion-10 {
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5rem;
  letter-spacing: normal;
  margin: 0;
  color: #151617;
}

.emotion-12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-item: center;
  -ms-flex-item-align-item: center;
  align-item: center;
  padding: 16px 32px;
  background-color: #636464;
}

.emotion-14 {
  border: 1px solid;
  text-transform: uppercase;
  font-weight: 600;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  padding: 12px 24px;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  cursor: pointer;
  width: 100%;
  min-width: 80px;
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 8px;
  color: #4F4F4F;
  background-color: #FFFFFF;
}

.emotion-14:disabled {
  color: #F8F9FA;
  background-color: #FFFFFF;
  border-color: #F8F9FA;
  cursor: default;
}

@media only screen and (min-width: 600px) {
  .emotion-14 {
    width: auto;
  }
}

.emotion-14:disabled {
  cursor: not-allowed;
}

.emotion-14:hover {
  color: #4F4F4F;
  background-color: #F8F9FA;
}

.emotion-16 {
  text-transform: uppercase;
  border: 0;
  font-weight: 600;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  padding: 12px 24px;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  cursor: pointer;
  width: 100%;
  min-width: 80px;
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 8px;
  color: #FFFFFF;
  background-color: #036DC1;
}

.emotion-16:focus {
  box-shadow: 0px 2px 1px 0px #1A1A1A0D,0px 0px 0px 4px #E6F5FF;
  background-color: #036DC1;
}

.emotion-16:disabled {
  opacity: 38%;
}

@media only screen and (min-width: 600px) {
  .emotion-16 {
    width: auto;
  }
}

.emotion-16:disabled {
  cursor: not-allowed;
}

.emotion-16:hover {
  color: #FFFFFF;
  background-color: #013C6B;
}

<div
  class="emotion-0 emotion-1"
  data-testid="modal-overlay"
>
  <div
    class="emotion-2 emotion-3"
  >
    <div
      class="emotion-4 emotion-5"
      data-testid="modal-header"
    >
      Title
      <button
        class="emotion-6 emotion-7"
        color="neutralDark_100"
        data-testid="icon-button"
      >
        <svg
          fill="none"
          height="12"
          viewBox="2 2 8 8"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.5 3.205L8.795 2.5L6 5.295L3.205 2.5L2.5 3.205L5.295 6L2.5 8.795L3.205 9.5L6 6.705L8.795 9.5L9.5 8.795L6.705 6L9.5 3.205Z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="emotion-8 emotion-9"
    >
      <p
        class="emotion-10 emotion-11"
        color="grey_300"
        data-testid="typography"
        style="margin-bottom: 8px; text-transform: none; font-weight: 700;"
      >
        Subtitle
      </p>
      Some text here
    </div>
    <div
      class="emotion-12 emotion-13"
    >
      <button
        class="emotion-14 emotion-15"
        color="neutral"
        data-testid="cancel-button"
      >
        Cancel
      </button>
      <button
        class="emotion-16 emotion-15"
        color="primary"
        data-testid="action-button"
      >
        Action
      </button>
    </div>
  </div>
</div>
`;
